libraries/AC_AutoTune/AC_AutoTune_Multi.h

```diff
@@ -0,0 +1,78 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/*
+  support for autotune of multirotors. Based on original autotune code from ArduCopter, written by Leonard Hall
+  Converted to a library by Andrew Tridgell
+ */
+
+#pragma once
+
+#include "AC_AutoTune.h"
+
+class AC_AutoTune_Multi : public AC_AutoTune
+{
+public:
+    // constructor
+    AC_AutoTune_Multi()
+    {
+        tune_seq[0] = 0;  // RD_UP
+        tune_seq[1] = 1; // RD_DOWN
+        tune_seq[2] = 2; // RP_UP
+        tune_seq[3] = 7; // SP_DOWN
+        tune_seq[4] = 6; // SP_UP
+        tune_seq[5] = 9; // tune complete
+    };
+    // save gained, called on disarm
+    void save_tuning_gains() override;
+
+protected:
+    void test_init() override;
+    void test_run(uint8_t test_axis, const float dir_sign) override;
+    void do_gcs_announcements() override;
+    void load_test_gains() override;
+    void updating_rate_p_up_all(uint8_t test_axis) override;
+    void updating_rate_p_down_all(uint8_t test_axis) override {};
+    void updating_rate_d_up_all(uint8_t test_axis) override;
+    void updating_rate_d_down_all(uint8_t test_axis) override;
+    void updating_angle_p_up_all(uint8_t test_axis) override;
+    void updating_angle_p_down_all(uint8_t test_axis) override;
+
+    void updating_rate_d_up(float &tune_d, float tune_d_min, float tune_d_max, float tune_d_step_ratio, float &tune_p, float tune_p_min, float tune_p_max, float tune_p_step_ratio, float rate_target, float meas_rate_min, float meas_rate_max);
```

Comment:
A few of these multi-only methods can probably be made private.  It seems like Heli will get some similar methods but they take different arguments so they're not really the same and can't be shared.

- updating_rate_d_up
- updating_rate_d_down
- updating_rate_p_up_d_down
- updating_angle_p_down
- updating_angle_p_up

Reply:
right.  I suppose these could be private methods.  These will be specific to the subclass.  Thanks!